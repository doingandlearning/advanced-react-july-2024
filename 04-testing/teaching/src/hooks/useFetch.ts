import { useState, useEffect } from "react";

export default function useFetch(url: string) {
  const [data, setData] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<null | Error>(null);

  useEffect(() => {
    let isMounted = true;
    const fetchData = async () => {
      try {
        setError(null);
        const response = await fetch(url);
        if (!response.ok) {
          throw new Error("We had a problem getting the data.");
        }
        const result = await response.json();
        if (isMounted) {
          setData(result);
          setLoading(false);
        }
      } catch (error) {
        if (isMounted) {
          setError(error as Error);
          setLoading(false);
        }
      }
    };

    fetchData();

    return () => {
      isMounted = false;
    };
  }, [url]);

  return { data, loading, error };
}
