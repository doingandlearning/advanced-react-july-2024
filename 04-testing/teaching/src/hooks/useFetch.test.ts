import { setupServer } from "msw/node";
import { http, HttpResponse } from "msw";
import { beforeAll, afterAll, afterEach, test, expect, describe } from "vitest";
import { renderHook, waitFor } from "@testing-library/react";
import useFetch from "./useFetch";
import { wait } from "@testing-library/user-event/dist/cjs/utils/index.js";

const server = setupServer(
  http.get("https://api.example.com/data", () => {
    return HttpResponse.json({ message: "Hello world" });
  })
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe("useFetch hook", () => {
  test("it returns data when the fetch is succesful", async () => {
    const { result } = renderHook(() =>
      useFetch("https://api.example.com/data")
    );

    expect(result.current.loading).toBe(true);
    expect(result.current.error).toBe(null);
    expect(result.current.data).toBe(null);

    await waitFor(() => {
      expect(result.current.loading).toBe(false);
      expect(result.current.error).toBe(null);
      expect(result.current.data).toEqual({ message: "Hello world" });
    });
  });

  test("it errors when the fetch is unsuccessful", async () => {
    server.use(
      http.get("https://api.example.com/data", () => {
        return new HttpResponse(null, {
          status: 404,
          statusText: "Out of apples",
        });
      })
    );
    const { result } = renderHook(() =>
      useFetch("https://api.example.com/data")
    );

    expect(result.current.loading).toBe(true);
    expect(result.current.error).toBe(null);
    expect(result.current.data).toBe(null);

    await waitFor(() => {
      expect(result.current.loading).toBe(false);
      expect(result.current.error).toMatchSnapshot();
      expect(result.current.data).toBeNull();
    });
  });
});
