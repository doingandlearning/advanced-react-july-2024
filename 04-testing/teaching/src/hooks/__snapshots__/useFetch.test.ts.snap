// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useFetch hook > it errors when the fetch is unsuccessful 1`] = `[Error: We had a problem getting the data.]`;
